### PR TITLE
Replace NotNullAttribute with thrown exceptions

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs [hunk 1]
```diff
@@ -14,8 +13,13 @@ public static class ConfigurationExtensions
         /// </summary>
         /// <param name="configuration"></param>
         /// <returns></returns>
-        public static IConfigurationBuilder AddUserSecrets([NotNull]this IConfigurationBuilder configuration)
+        public static IConfigurationBuilder AddUserSecrets(this IConfigurationBuilder configuration)
         {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
```
Comment (muratg) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: Does `nameof` work for `this` parameters in the extension methods?

**Comment by muratg** [comment]
Once the travis is green, :shipit:

File: src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: The parameter name isn't `this`, it's `configuration`. The `this` is just syntactic sugar for the compiler.

Comment (muratg) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: @Eilon Yup I know. I wasn't sure if `nameof(configuration)` would be equal to `"configuration"` or something else.  Not sure how it works in extensions method with the syntactic sugar.  Even if it's `"configuration"` I'm not sure if it's very useful in the error message, or if just throwing a `NullReferenceException` would be a better choice for when `this` is null in extension methods.

Comment (Eilon) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: Disagree. The parameter name is `configuration` no matter how you look at it. A `NullReferenceException` gives the developer no indication about which parameter is invalid. Extension methods are just plain old static methods.

Comment (muratg) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: Agreed, but name `configuration` is not visible to the user of the extension method. To the user, it mostly looks like a normal function (except the `(extension)` text in front of the signature): 

image

Comment (Eilon) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: Press <kbd>F12</kbd> on the method :smile:

Comment (pranavkm) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: Where did we land on this?

Comment (muratg) [inline src/Microsoft.Extensions.Configuration.UserSecrets/ConfigurationExtensions.cs]: Let's go with the `ArgumentNullException`.  If someone hits this and can't navigate to the root cause, we'll explain :)

**Pull request merged**